opte/src/engine/port.rs

```diff
@@ -1059,13 +1026,44 @@ impl Port {
 
         Err(OpteError::LayerNotFound(layer_name.to_string()))
     }
+
+    /// Grab a snapshot of the port statistics.
+    pub fn stats_snap(&self) -> PortStatsSnap {
+        self.data.lock().stats.vals.snapshot()
```

Comment:
The locking here makes me a bit worried about deadlocks. I might at least put something in the comment that this acquires the lock and can't be called from any other function that also does so. AFAICT, this is only used in the tests to check that the stats behave as we expect, but still might be a good idea.

Reply:
Good news, in the kernel recursive mutex entries panic the box! So no worry about deadlocks. This is kind of a problem in general with all the Port methods, you have to know if it's expecting to grab the lock or already hold it. I achieved this my having some methods take a `&mut PortData`, which is only possible if you hold the lock. Of course, this doesn't stop you from calling another method on `self` that might also acquire said lock. This might be an argument for making "internal" methods be static so that `self` and `&mut PortData` cannot coexist. Or perhaps moving some methods on `PortData` instead. Until then, I'm just going to put this behind test cfg, because that's all it's used for currently. If some userland process needs these stats they can use kstats. I'll also make a note about the locking and write up an issue.

The good news is, practically speaking, we'll hit this in dev immediately by seeing a panic. So it's also worth not over rotating on this. The main way it could bite is if we have some API that is rarely used and has a recursive entry. Which gives me another thought: can we configure the std Mutex impl to panic on recursive entry? If so, we could also catch this when running tests. If it _doesn't_ support that feature, perhaps we can implement the `std` version of KMutex with our own direct bindings to the pthread mutex.

Reply:
Sounds good, thanks for that.

> can we configure the std Mutex impl to panic on recursive entry?

Nope, I just ran into this last week. There is no support for configuring what happens in this case, and `std::sync::Mutex` is set up to just deadlock. Bummer. We may want to do make our own mutex over pthreads that's designed to panic here too, if we want parity between the userland and kernel implementations.